ambry-router/src/main/java/com/github/ambry/router/QuotaAwareOperationController.java

```diff
@@ -0,0 +1,193 @@
+    try {
+      List<RequestInfo> newRequestsToSend = new ArrayList<>();
+      pollNewRequests(newRequestsToSend, requestsToDrop);
+      addToRequestQueue(newRequestsToSend);
+      drainRequestQueue(requestsToSend);
+    } catch (Exception e) {
+      logger.error("Operation Manager poll received an unexpected error: ", e);
+      routerMetrics.operationManagerPollErrorCount.inc();
+    }
+  }
+
+  /**
+   * Poll for new requests from each of the operation managers, and initiate backgroud deletes.
+   * @param requestsToSend a list of {@link RequestInfo} that will contain the new requests to be sent out.
+   * @param requestsToDrop a list of correlation IDs that will contain the IDs for requests that the network layer
+   *                       should drop.
+   */
+  private void pollNewRequests(List<RequestInfo> requestsToSend, Set<Integer> requestsToDrop) {
+    putManager.poll(requestsToSend, requestsToDrop);
+    getManager.poll(requestsToSend, requestsToDrop);
+    nonBlockingRouter.initiateBackgroundDeletes(getBackGroundDeleteRequests());
+    clearBackGroundDeleteRequests();
+    deleteManager.poll(requestsToSend, requestsToDrop);
+    ttlUpdateManager.poll(requestsToSend, requestsToDrop);
+    undeleteManager.poll(requestsToSend, requestsToDrop);
+  }
+
+  /**
+   * Add the specified {@link List} of {@link RequestInfo}s to the reqeust queue.
+   * @param requestInfos {@link List} of {@link RequestInfo} objects.
+   */
+  private void addToRequestQueue(List<RequestInfo> requestInfos) {
+    for (RequestInfo requestInfo : requestInfos) {
+      QuotaResource quotaResource = requestInfo.getChargeable().getQuotaResource();
+      if (quotaResource == null) {
+        quotaResource = UNKNOWN_QUOTA_RESOURCE;
+      }
+      requestQueue.putIfAbsent(quotaResource, new LinkedList<>());
+      requestQueue.get(quotaResource).add(requestInfo);
+    }
+  }
+
+  /**
+   * Drain the request queue based on quota and system resources and update the requests to be send to {@code requestsToSend}.
+   * @param requestsToSend a list of {@link RequestInfo} that will contain the requests to be sent out.
+   */
+  private void drainRequestQueue(List<RequestInfo> requestsToSend) {
+    drainQuotaCompliantRequests(requestsToSend);
+
+    List<QuotaResource> quotaResources = new ArrayList<>(requestQueue.keySet());
+    Collections.shuffle(quotaResources);
+    boolean done = false;
+    while (!requestQueue.isEmpty()) {
+      for (QuotaResource quotaResource : quotaResources) {
+        RequestInfo requestInfo = requestQueue.get(quotaResource).getFirst();
+        if (requestInfo.getChargeable().quotaExceedAllowed()) {
+          requestInfo.getChargeable().charge();
+          requestsToSend.add(requestInfo);
+          requestQueue.get(quotaResource).removeFirst();
+          if (requestQueue.get(quotaResource).isEmpty()) {
+            requestQueue.remove(quotaResource);
+          }
+        } else {
+          done = true;
+          break;
```

Comment:
This is weird. Your exiting on the first request returning false for quotaExceedAllowed() implies that quotaExceedAllowed() is actually not a property of the request but rather a global setting. You should use a static class to access that setting and either completely skip the list or go through the all items.

Reply:
You are right about this. I realized that quotaExceedAllowed() needs to be per request because for quota purposes there are two types of requests
- read requests that charge RCU
- write requests that charge WCU and Storage quota.
Frontend node accounts differently for WCU capacity and RCU capacity.